packages/airnode-protocol/contracts/rrp/authorizers/admin/AirnodeFeeRegistry.sol

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+import "./AirnodeFeeRegistryRolesWithManager.sol";
+import "./interfaces/IAirnodeFeeRegistry.sol";
+
+/// @title Contract for storing the price of a airnode endpoints
+/// @notice The AirnodeFeeRegistry contract is a central contract that
+/// all the different AirnodeTokenLock contracts (of different chains)
+/// will be calling to find out what the USD price of a chain-airnode-endpoint is
+contract AirnodeFeeRegistry is
+    AirnodeFeeRegistryRolesWithManager,
+    IAirnodeFeeRegistry
+{
+    /// @dev The USD price of an endpoint for an airnode on a specific chain
+    mapping(uint256 => mapping(address => mapping(bytes32 => uint256)))
+        public chainIdToAirnodeToEndpointToPrice;
+
+    /// @dev The USD price of an endpoint for an airnode across all chains
+    mapping(address => mapping(bytes32 => uint256))
+        public airnodeToEndpointToPrice;
+
+    /// @dev The default price of an endpoint for an airnode on a specific chain
+    mapping(uint256 => mapping(address => uint256))
+        public defaultChainAirnodePrice;
+
+    /// @dev The default price of an endpoint on an airnode
+    mapping(address => uint256) public defaultAirnodePrice;
+
+    /// @dev The default price of an endpoint for a chain
+    mapping(uint256 => uint256) public defaultChainPrice;
+
+    /// @dev The global default price of an endpoint
+    uint256 public defaultPrice;
```

Comment:
How do you feel about replacing all these variables for a single mapping using a data structures like the this below:

```
    struct EndpointPrice {
        uint256 price;
    }

    struct AirnodePrice {
        uint256 defaultPrice;
        uint256 price;
        mapping(bytes32 => EndpointPrice) endpointPrices;
    }

    struct ChainPrice {
        uint256 defaultPrice;
        uint256 price;
        mapping(address => AirnodePrice) airnodePrices;
    }

    mapping(uint256 => ChainPrice) public chainIdToAirnodeToEndpointToPrice;
``` 

The `getEndpointPrice()` function would be something like this:
```
    function getEndpointPrice(
        uint256 _chainId,
        address _airnode,
        bytes32 _endpointId
    ) external view returns (uint256) {
        if (
            chainIdToAirnodeToEndpointToPrice[_chainId]
                .airnodePrices[_airnode]
                .endpointPrices[_endpointId]
                .price != 0
        )
            // returns the endpoint price for an airnode in a chain
            return
                chainIdToAirnodeToEndpointToPrice[_chainId]
                    .airnodePrices[_airnode]
                    .endpointPrices[_endpointId]
                    .price;

        if (
            airnodeEndpointFlag[_airnode] &&
            chainIdToAirnodeToEndpointToPrice[0]
                .airnodePrices[_airnode]
                .endpointPrices[_endpointId]
                .price !=
            0
        )
            // returns the endpoint price for an airnode across all chains
            return
                chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .endpointPrices[_endpointId]
                    .price;
        // returns the default price for an airnode in a chain
        else
            return
                chainIdToAirnodeToEndpointToPrice[_chainId]
                    .airnodePrices[_airnode]
                    .defaultPrice;

        if (
            chainIdToAirnodeToEndpointToPrice[0]
                .airnodePrices[_airnode]
                .defaultPrice != 0
        )
            // returns the default price for an airnode across all chains
            return
                chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .defaultPrice;
        if (chainIdToAirnodeToEndpointToPrice[_chainId].defaultPrice != 0)
            // returns the default price for a chain
            return chainIdToAirnodeToEndpointToPrice[_chainId].defaultPrice;

        // returns the default price for all airnodes across all chains
        return
            chainIdToAirnodeToEndpointToPrice[0]
                .airnodePrices[address(0)]
                .endpointPrices[bytes32(0)]
                .price;
    }
```

Reply:
This was something I thought of doing initially, this could be made even simpler if we treat the [0] indexes as the default price so something like this

```
       function getEndpointPrice(
        uint256 _chainId,
        address _airnode,
        bytes32 _endpointId
    ) external view returns(uint256) {
        if (
            chainIdToAirnodeToEndpointToPrice[_chainId][_airnode][_endpointId]
            != 0
        )
            // returns the endpoint price for an airnode in a chain
            return
                chainIdToAirnodeToEndpointToPrice[_chainId][_airnode][_endpointId]
        if (
            airnodeEndpointFlag[_airnode] &&
            chainIdToAirnodeToEndpointToPrice[0][_airnode][_endpointId] != 0
        )
            // returns the endpoint price for an airnode across all chains
            return
                chainIdToAirnodeToEndpointToPrice[0][_airnode][_endpointId]
            // returns the default price for an airnode in a chain
        else
            return
                chainIdToAirnodeToEndpointToPrice[_chainId][_airnode][0];

        if (
            chainIdToAirnodeToEndpointToPrice[0][_airnode][0] != 0
        )
            // returns the default price for an airnode across all chains
            return
                chainIdToAirnodeToEndpointToPrice[0][_airnode][0];
        if (chainIdToAirnodeToEndpointToPrice[_chainId][0][0] != 0)
            // returns the default price for a chain
            return 
                chainIdToAirnodeToEndpointToPrice[_chainId][0][0];

        // returns the default price for all airnodes across all chains
        return
            chainIdToAirnodeToEndpointToPrice[0][0][0]
    }

```

I agree I will probably change it to a variation of the above, one thing to keep in mind is that endpointId can be 0(correct me if I am wrong on this) so thats why I was hesitant to go with this approach, but your approach makes it easier, I would just return a global default price similar to now and keep everything else the same.

Reply:
So this was the what I ended up with 

```
struct EndpointPrice {
        uint256 price;
    }

    struct AirnodePrice {
        uint256 defaultPrice;
        mapping(bytes32 => EndpointPrice) endpointPrices;
    }

    struct ChainPrice {
        uint256 defaultPrice;
        mapping(address => AirnodePrice) airnodePrices;
    }

    /// @dev A mapping structure used to store all the prices for all the different chain-airnode-endpoint pairs.
    /// Default prices are also stored for the respective chains,airnodes. 
    mapping(uint256 => ChainPrice) public chainIdToAirnodeToEndpointToPrice;

    /// @dev The global default price of an endpoint
    uint256 public defaultPrice;
    
    function getEndpointPrice(
        uint256 _chainId,
        address _airnode,
        bytes32 _endpointId
    ) external view override returns (uint256) {
        if (
            chainIdToAirnodeToEndpointToPrice[_chainId]
                .airnodePrices[_airnode]
                .endpointPrices[_endpointId]
                .price != 0
        )
            // returns the endpoint price for an airnode in a chain
            return
                chainIdToAirnodeToEndpointToPrice[_chainId]
                .airnodePrices[_airnode]
                .endpointPrices[_endpointId]
                .price;

        if (airnodeEndpointFlag[_airnode]) {

            if (
                chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .endpointPrices[_endpointId]
                    .price != 0
            )
                // returns the endpoint price for an airnode across all chains
                return
                    chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .endpointPrices[_endpointId]
                    .price;

            
            if (
                chainIdToAirnodeToEndpointToPrice[_chainId]
                    .airnodePrices[_airnode]
                    .defaultPrice != 0
            )
                // returns the default price of an endpoint on an airnode on a chain
                return
                    chainIdToAirnodeToEndpointToPrice[_chainId]
                        .airnodePrices[_airnode]
                        .defaultPrice;
        
        } else {

            if (
                chainIdToAirnodeToEndpointToPrice[_chainId]
                    .airnodePrices[_airnode]
                    .defaultPrice != 0
            )
                // returns the default price of an endpoint on an airnode on a chain
                return
                    chainIdToAirnodeToEndpointToPrice[_chainId]
                        .airnodePrices[_airnode]
                        .defaultPrice;
            if (
                chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .endpointPrices[_endpointId]
                    .price != 0
            )
                // returns the endpoint price for an airnode across all chains
                return
                    chainIdToAirnodeToEndpointToPrice[_chainId]
                        .airnodePrices[_airnode]
                        .endpointPrices[_endpointId]
                        .price;
        }

        if (
            chainIdToAirnodeToEndpointToPrice[0]
                .airnodePrices[_airnode]
                .defaultPrice != 0
        )
         // returns the default price for an airnode across all chains
            return 
                chainIdToAirnodeToEndpointToPrice[0]
                    .airnodePrices[_airnode]
                    .defaultPrice;
                    
        if (chainIdToAirnodeToEndpointToPrice[_chainId].defaultPrice != 0)
        // returns the default price for a chain
            return chainIdToAirnodeToEndpointToPrice[_chainId].defaultPrice;
        // returns the global default price
        return defaultPrice;
    }
```

Which definitely looks a bit cleaner but there is a problem with this approach in that the mapping inside the structs don't get initialized with [the normal getters](https://docs.soliditylang.org/en/v0.5.7/contracts.html#visibility-and-getters), this means we would have to write our own getters to access the mappings for us to be able to test the values inside. Just felt like a longer workaround than the old way. One other way is how I described above with 0 indexes being treated as default prices

Reply:
Makes sense, thanks for trying it out anyways 👏🏻 

I think it's fine to go back to the old way since my suggestion didn't really improve the number of txs needed to set this whole structure of fees and storage savings might not be impacted that much. Also, I didn't consider the endpointId possibly being 0.